File: packages/python/procedurals/stop_ramp.py
"""Immutable noncyclic positioned RGB24 color stops independently specified by the
shared color.stop-ramp contract.

Motivating notes: survey/out/2016/Generativos/boxDepth/notes.md,
survey/out/2016/Generativos/celular/notes.md,
survey/out/2016/Generativos/colorRamp/notes.md and
survey/out/2016/Generativos/triangleRamp/notes.md. No artistic stop-count/position/
palette range is established; see the catalog contract for evidence.
"""
from dataclasses import dataclass
import math

__all__ = ["StopRampError", "stop_ramp"]
_MAX_COLOR = 16777215


class StopRampError(ValueError):
    """Stable catalog code: INVALID_INPUT or INVALID_QUERY."""
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def _fail(code):
    raise StopRampError(code)


def _number(value, code):
    if type(value) not in (int, float) or type(value) is bool:
        _fail(code)
    try:
        value = float(value)
    except OverflowError:
        raise StopRampError(code) from None
    if not math.isfinite(value):
        _fail(code)
    return value


def _checked_position(value):
    position = _number(value, "INVALID_INPUT")
    if position < 0 or position > 1:
        _fail("INVALID_INPUT")
    return 0.0 if position == 0 else position


def _checked_color(value):
    color = _number(value, "INVALID_INPUT")
    if color < 0 or color > _MAX_COLOR or color != math.floor(color):
        _fail("INVALID_INPUT")
    return int(color)


def _channel(a, b, t):
    difference = b - a
    product = difference * t
    value = a + product
    return math.floor(value + 0.5)


@dataclass(frozen=True, slots=True)
class _Ramp:
    _positions: list
    _colors: list

    def serialize(self):
        """Export detached ordered position/color stop records; retain none."""
        return {"stops": [{"position": p, "color": c} for p, c in zip(self._positions, self._colors)]}

    def sample(self, query):
        """Sample one RGB24 color; endpoint holds outside the first/last stop."""
        value = _number(query, "INVALID_QUERY")
        positions, colors = self._positions, self._colors
        if value <= positions[0]:
            return colors[0]
        last = len(positions) - 1
        if value >= positions[last]:
            return colors[last]
        left, right = 0, last
        while right - left > 1:
            middle = left + (right - left) // 2
            if value < positions[middle]:
                right = middle
            else:
                left = middle
        if value == positions[left]:
            return colors[left]
        if value == positions[right]:
            return colors[right]
        numerator = value - positions[left]
        denominator = positions[right] - positions[left]
        t = numerator / denominator
        a, b = colors[left], colors[right]
        red = _channel((a >> 16) & 255, (b >> 16) & 255, t)
        green = _channel((a >> 8) & 255, (b >> 8) & 255, t)
        blue = _channel(a & 255, b & 255, t)
        return red * 65536 + green * 256 + blue


def stop_ramp(config):
    """Build immutable positioned RGB24 color stops from {"stops": [{"position","color"}, ...]}.

    Implements color.stop-ramp 0.1.0 independently of source code.
    """
    if type(config) is not dict or set(config) != {"stops"}:
        _fail("INVALID_INPUT")
    stop_values = config["stops"]
    if type(stop_values) is not list:
        _fail("INVALID_INPUT")
    if len(stop_values) == 0:
        _fail("INVALID_INPUT")
    positions = [0.0] * len(stop_values)
    colors = [0] * len(stop_values)
    for index, stop_value in enumerate(stop_values):
        if type(stop_value) is not dict or set(stop_value) != {"position", "color"}:
            _fail("INVALID_INPUT")
        position = _checked_position(stop_value["position"])
        if index > 0 and not (position > positions[index - 1]):
            _fail("INVALID_INPUT")
        positions[index] = position
        colors[index] = _checked_color(stop_value["color"])
    return _Ramp(positions, colors)

File: packages/javascript/src/stop-ramp.js
const MAX_COLOR = 16777215;

/** Error with one of the stable color.stop-ramp contract codes. */
export class StopRampError extends Error {
  constructor(code) { super(code); this.name = "StopRampError"; this.code = code; }
}

function fail(code) { throw new StopRampError(code); }
function finite(value) { return typeof value === "number" && Number.isFinite(value); }
function number(value, code) { if (!finite(value)) fail(code); return value; }

function passiveRecord(value, keys, code) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail(code);
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) fail(code);
  if (Reflect.ownKeys(value).length !== keys.length) fail(code);
  for (const key of keys) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !("value" in descriptor)) fail(code);
  }
  return value;
}
function recordValue(record, key) { return Object.getOwnPropertyDescriptor(record, key).value; }
function passiveArray(value, code) {
  if (!Array.isArray(value) || Object.getPrototypeOf(value) !== Array.prototype) fail(code);
  return value;
}
function arrayItem(array, index, code) {
  const descriptor = Object.getOwnPropertyDescriptor(array, String(index));
  if (descriptor === undefined || !("value" in descriptor)) fail(code);
  return descriptor.value;
}

function checkedPosition(value) {
  const position = number(value, "INVALID_INPUT");
  if (position < 0 || position > 1) fail("INVALID_INPUT");
  return position === 0 ? 0 : position;
}
function checkedColor(value) {
  const color = number(value, "INVALID_INPUT");
  if (color < 0 || color > MAX_COLOR || color !== Math.floor(color)) fail("INVALID_INPUT");
  return color;
}
function channel(a, b, t) {
  const difference = b - a;
  const product = difference * t;
  const value = a + product;
  return Math.floor(value + 0.5);
}

class Ramp {
  #positions; #colors;
  constructor(positions, colors) {
    this.#positions = positions;
    this.#colors = colors;
    Object.freeze(this);
  }
  serialize() {
    const stops = [];
    for (let i = 0; i < this.#positions.length; i += 1) {
      stops.push({ position: this.#positions[i], color: this.#colors[i] });
    }
    return { stops };
  }
  sample(query) {
    const value = number(query, "INVALID_QUERY");
    const positions = this.#positions, colors = this.#colors;
    if (value <= positions[0]) return colors[0];
    const last = positions.length - 1;
    if (value >= positions[last]) return colors[last];
    let left = 0, right = last;
    while (right - left > 1) {
      const middle = left + Math.floor((right - left) / 2);
      if (value < positions[middle]) right = middle; else left = middle;
    }
    if (value === positions[left]) return colors[left];
    if (value === positions[right]) return colors[right];
    const numerator = value - positions[left];
    const denominator = positions[right] - positions[left];
    const t = numerator / denominator;
    const a = colors[left], b = colors[right];
    const red = channel((a >>> 16) & 255, (b >>> 16) & 255, t);
    const green = channel((a >>> 8) & 255, (b >>> 8) & 255, t);
    const blue = channel(a & 255, b & 255, t);
    return red * 65536 + green * 256 + blue;
  }
}

/**
 * Immutable noncyclic positioned RGB24 color stops with piecewise linear sampling and
 * endpoint holds. Implements color.stop-ramp 0.1.0 independently of source code.
 * Motivating sketches: survey/out/2016/Generativos/boxDepth, celular, colorRamp and
 * triangleRamp; no artistic stop-count/position/palette range is established, see the
 * catalog contract for evidence.
 */
export function stopRamp(input) {
  const record = passiveRecord(input, ["stops"], "INVALID_INPUT");
  const stopValues = passiveArray(recordValue(record, "stops"), "INVALID_INPUT");
  if (stopValues.length === 0) fail("INVALID_INPUT");
  const positions = new Array(stopValues.length), colors = new Array(stopValues.length);
  for (let index = 0; index < stopValues.length; index += 1) {
    const stopRecord = passiveRecord(arrayItem(stopValues, index, "INVALID_INPUT"),
      ["position", "color"], "INVALID_INPUT");
    const position = checkedPosition(recordValue(stopRecord, "position"));
    if (index > 0 && !(position > positions[index - 1])) fail("INVALID_INPUT");
    positions[index] = position;
    colors[index] = checkedColor(recordValue(stopRecord, "color"));
  }
  return new Ramp(positions, colors);
}
